lib/api/api.go

```diff
@@ -105,7 +108,14 @@ type Service interface {
 	WaitForStart() error
 }
 
-func New(id protocol.DeviceID, cfg config.Wrapper, assetDir, tlsDefaultCommonName string, m model.Model, defaultSub, diskSub events.BufferedSubscription, evLogger events.Logger, discoverer discover.Manager, connectionsService connections.Service, urService *ur.Service, fss model.FolderSummaryService, errors, systemLog logger.Recorder, noUpgrade bool, miscDB *db.NamespacedKV) Service {
+func New(id protocol.DeviceID, cfg config.Wrapper, assetDir, tlsDefaultCommonName string, m model.Model, defaultSub, diskSub events.BufferedSubscription, evLogger events.Logger, discoverer discover.Manager, connectionsService connections.Service, urService *ur.Service, fss model.FolderSummaryService, errors, systemLog logger.Recorder, noUpgrade bool, miscDB *db.NamespacedKV) (Service, error) {
+
+	tokenCookieManager := newTokenCookieManager(id.Short().String(), cfg.GUI(), evLogger, miscDB)
+	webauthnService, err := newWebauthnService(tokenCookieManager, cfg, evLogger, miscDB, "webauthn")
```

Comment:
There's an issue regarding config changes and service lifetimes here. I can very much see how it happened, both the `config.Wrapper` interface in the first place, and our usage in `lib/api` (and other places) is quite a mess in regards to the issue I am pointing out below:

I tried to explain it generally, but I am not sure if this is helpful/comprehensible - I'll explain the concrete issue here below which should hopefully be clear:  
An object is either a service, meaning it's registered with the suture service handler and with the config wrapper to get notified of config changes (if needed), or it's part of a service which takes care of lifetime and config changes.

Previously `tokenCookieManager` was constructed in `api.service.Serve`. And that service restarts on config changes, thus the token manager gets constructed again with the new config when `Serve` is called again. Now it will never notice the config changed.

As for the new webauthn service: It gets the wrapper, however internally does a mix of extracting info from the config at construction time and calling the wrapper directly. Calling the wrapper may result in more correct behaviour, as config changes are reflected, however it's not the way to do it still for a few reasons: E.g. the overall state of the webauthn service may be inconsistent, due to timing issues when calling wrapper methods.

I don't see why the token manager needs to be a field on the struct, afaik all usages are in `Serve`. Resp. that one handler function, however you can still initialize that from `Serve` passing the token manager to a function that creates the handler. So afaik the simplest fix is to move construction of both these back to `Serve`, just passing the needed config types (not wrapper) to the constructors.

Reply:
Thanks for explaining! I think I see what you mean, that does seem to make sense. I think another reason I moved these instantiations around was that I previously needed access to the WebAuthn state manager in the config manager, but that is also obsoleted by the fix for https://github.com/syncthing/syncthing/pull/9175#discussion_r1700853007. So yeah, I think these can be disentangled like you suggest.

Reply:
Should be fixed by https://github.com/emlun/syncthing/pull/4, from what I can tell this fully reverts the changes to the construction of `tokenCookieManager`.